### ISSUE TITLE
`unused_async_trait_impl` is not working well with `!`

### ISSUE THREAD

**Issue by nazar-pc** [rust-lang/rust-clippy#17162]
### Summary

`unused_async_trait_impl` should probably recognize `!` type and not trigger on it

### Lint Name

unused_async_trait_impl

### Reproducer

I tried this code:

```rust
impl ReadAtAsync for ! {
    async fn read_at<B>(&self, _buf: B, _offset: u64) -> io::Result<B>
    where
        AsyncReadBytes<B>: From<B>,
        B: AsMut<[u8]> + Unpin + 'static,
    {
        unreachable!("Is never called")
    }
}
```

I saw this happen:

```
warning: unused `async` for async trait impl function with no `.await` statements
   --> crates/farmer/ab-farmer-components/src/lib.rs:150:5
    |
150 | /     async fn read_at<B>(&self, _buf: B, _offset: u64) -> io::Result<B>
151 | |     where
152 | |         AsyncReadBytes<B>: From<B>,
153 | |         B: AsMut<[u8]> + Unpin + 'static,
154 | |     {
155 | |         unreachable!("Is never called")
156 | |     }
    | |_____^
    |
    = note: `std::future::ready` creates a `Future` which returns the value immediately when `poll`ed
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unused_async_trait_impl
    = note: `-W clippy::unused-async-trait-impl` implied by `-W clippy::pedantic`
    = help: to override `-W clippy::pedantic` add `#[allow(clippy::unused_async_trait_impl)]`
help: consider removing the `async` from this function and returning `impl Future<Output = io::Result<B>>` instead
    |
150 ~     fn read_at<B>(&self, _buf: B, _offset: u64) -> impl Future<Output = io::Result<B>>
151 |     where
...
154 |     {
155 ~         std::future::ready(unreachable!("Is never called"))
    |
```

Applying the suggestion leads to a different warning though:
```
   --> crates/farmer/ab-farmer-components/src/lib.rs:155:9
    |
155 |         ready(unreachable!("Is never called"))
    |         ^^^^^ ------------------------------- any code following this expression is unreachable
    |         |
    |         unreachable call
    |
    = note: `#[warn(unreachable_code)]` (part of `#[warn(unused)]`) on by default
```

I think it'd be better to special-case `!` and not trigger this lint on it at all.

### Version

```text
rustc 1.98.0-nightly (e7815e522 2026-06-04)
binary: rustc
commit-hash: e7815e522ecc746592fee32f50478f521333b503
commit-date: 2026-06-04
host: x86_64-unknown-linux-gnu
release: 1.98.0-nightly
LLVM version: 22.1.6
```

**Comment by ada4a** [comment]
> I think it'd be better to special-case `!` and not trigger this lint on it at all.

Hm, I _think_ avoiding `async fn` still saves you some compilation time by not constructing the state machine. So maybe the lint should still trigger, but suggest just `unreachable!()` instead of `ready(unreachable!())`?

cc @Wassasin as the author of the lint, what do you think?

**Comment by nazar-pc** [comment]
> So maybe the lint should still trigger, but suggest just `unreachable!()` instead of `ready(unreachable!())`?

That will not compile at all in today's Rust:
```
error[E0277]: `!` is not a future
   --> crates/farmer/ab-farmer-components/src/lib.rs:154:52
    |
154 |     fn read_at<B>(&self, _buf: B, _offset: u64) -> impl Future<Output = io::Result<B>>
    |                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `!` is not a future
...
159 |         unreachable!("Is never called")
    |         ------------------------------- return type was inferred to be `!` here
    |
    = help: the trait `futures::Future` is not implemented for `!`
```

**Comment by ada4a** [comment]
Huh!! Makes sense though, if you think about it. So I guess the solution is indeed to not lint in this case. Very curious edge case you've found here, thanks for reporting!

**Comment by Wassasin** [comment]
I think two separate things are happening here:
* this lint & suggested fix are redundant for implementations on `!` (never).
* only bodies containing macro invocations of `todo!` and `unimplemented!` are ignored. However all expressions with non-inhabited types will result in an `unreachable` lint for the wrapped `core::futures::ready` expression. Perhaps we should ignore bodies of which the final expression are a non-inhabited type from emitting this lint.

The following also yields the not very useful suggestion:
```rust
struct Test;

impl crate::HasAsyncMethod for Test {
    async fn do_something() -> u32 {
        //~^ unused_async_trait_impl
        panic!("test")
    }
}
```

I agree we should fix the latter. However with regards to the never-type, the lint is not necessarily incorrect, just redundant as the methods can never be called (and hence will not result in generated code, making the issue moot).
